xarray/core/computation.py

```diff
@@ -697,16 +711,26 @@ def stack(objects, dim, new_coord):
         this_apply = functools.partial(
             apply_ufunc, func, signature=signature, join=join,
             dask_array=dask_array, exclude_dims=exclude_dims,
-            dataset_fill_value=dataset_fill_value)
+            dataset_fill_value=dataset_fill_value,
+            data_vars_join=data_vars_join)
         return apply_groupby_ufunc(this_apply, *args)
     elif any(is_dict_like(a) for a in args):
-        return apply_dataset_ufunc(variables_ufunc, *args, signature=signature,
-                                   join=join, exclude_dims=exclude_dims,
-                                   fill_value=dataset_fill_value)
+        first_obj = args[0]
+        out = apply_dataset_ufunc(variables_ufunc, *args, signature=signature,
+                                  join=join, exclude_dims=exclude_dims,
+                                  fill_value=dataset_fill_value,
+                                  data_vars_join=data_vars_join)
+        if keep_attrs and (type(first_obj) is type(out)):
```

Comment:
I would move `keep_attrs` one level up into `apply_dataset_ufunc` and `apply_dataarray_ufunc`.